source/client.c

```diff
@@ -175,38 +193,45 @@ static void s_mqtt_client_shutdown(
                 (void *)connection);
             MQTT_CLIENT_CALL_CALLBACK(connection, on_disconnect);
             break;
-        case AWS_MQTT_CLIENT_STATE_CONNECTING:
-            AWS_LOGF_TRACE(
-                AWS_LS_MQTT_CLIENT, "id=%p: Initial connection attempt failed, calling callback", (void *)connection);
-            MQTT_CLIENT_CALL_CALLBACK_ARGS(connection, on_connection_complete, error_code, 0, false);
-            break;
-        case AWS_MQTT_CLIENT_STATE_CONNECTED: {
-            AWS_LOGF_DEBUG(
-                AWS_LS_MQTT_CLIENT,
-                "id=%p: Connection interrupted, calling callback and attempting reconnect",
-                (void *)connection);
-            MQTT_CLIENT_CALL_CALLBACK_ARGS(connection, on_interrupted, error_code);
 
-            /* In case user called disconnect from the on_interrupted callback */
-            bool stop_reconnect;
-            { /* BEGIN CRITICAL SECTION */
-                mqtt_connection_lock_synced_data(connection);
-                stop_reconnect = connection->synced_data.state == AWS_MQTT_CLIENT_STATE_DISCONNECTING;
+        case AWS_MQTT_CLIENT_STATE_CONNECTED: {
+            if (connection->auto_reconnect) {
+                AWS_LOGF_DEBUG(
+                    AWS_LS_MQTT_CLIENT,
+                    "id=%p: Connection interrupted, calling callback and attempting reconnect",
+                    (void *)connection);
+                MQTT_CLIENT_CALL_CALLBACK_ARGS(connection, on_interrupted, error_code);
+
+                /* In case user called disconnect from the on_interrupted callback */
+                bool stop_reconnect;
+                { /* BEGIN CRITICAL SECTION */
+                    mqtt_connection_lock_synced_data(connection);
+                    stop_reconnect = connection->synced_data.state == AWS_MQTT_CLIENT_STATE_DISCONNECTING;
```

Comment:
Should we also set `connection->auto_reconnect` to `false` somewhere in here?

Reply:
In this case, there will be no further reconnect happening for this connection object. To reconnect/connect, user will need to call `aws_mqtt_client_connection_connect` with new option. So, it's not necessary to change the `connection->auto_reconnect` here.